crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -414,7 +414,40 @@ where
         &self,
         number: BlockNumber,
     ) -> ProviderResult<Option<StoredBlockBodyIndices>> {
-        self.database.block_body_indices(number)
+        if let Some(indices) = self.database.block_body_indices(number)? {
+            Ok(Some(indices))
+        } else if self.canonical_in_memory_state.hash_by_number(number).is_some() {
```

Comment:
this should also use the `BlockState` and traverse its parent blocks for this and the anchor block number to read the indices from disk